File: llmc/state.py
"""Proxy state - persisted to the llmc-state bind path.

The proxy keeps a small amount of mutable state on disk so that:
    1. A proxy restart recovers the active mode + model without an
       expensive "what's running?" GPU service probe
    2. The CLI can read the state directly (via `llmc status`) without
       hitting the proxy HTTP endpoint
    3. Operators can inspect / reset state with `cat` / `rm`

State lives at /state/active.toml inside the proxy container - bind-mounted
from $HOME/docker-volumes/state on the host (see volumes.toml). Schema:

    mode = "llm"            # "llm" | "comfyui" | "train" | "idle"
    model = "qwen36"        # Active preset name (only set when mode = "llm")
    lock_owners = ["a", "b"] # List of owners holding the lock
    updated_at = 1747162456 # Unix timestamp of last update

Writes are atomic via temp-file + rename. Both fields are optional in
the on-disk file; the loader fills in defaults.
"""

from __future__ import annotations

import os
import time
import tomllib
import json
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import Optional

MODES = frozenset({"llm", "comfyui", "train", "idle"})


class StateError(ValueError):
    """Raised when state file is malformed."""


@dataclass(frozen=True)
class State:
    mode: str = "idle"
    model: Optional[str] = None
    lock_owners: list[str] = field(default_factory=list)
    updated_at: int = 0

    def __post_init__(self):
        if self.mode not in MODES:
            raise StateError(f"invalid mode {self.mode!r}; must be one of {sorted(MODES)}")
        if self.mode == "llm" and not self.model:
            # Allowed: someone is restarting the proxy with no model yet.
            # Logged by the caller, not an error.
            pass

    def to_toml(self) -> str:
        lines = [f'mode = "{self.mode}"']
        if self.model is not None:
            lines.append(f'model = "{self.model}"')
        if self.lock_owners:
            lines.append(f'lock_owners = {json.dumps(self.lock_owners)}')
        lines.append(f"updated_at = {self.updated_at}")
        return "\n".join(lines) + "\n"


def load(path: Path) -> State:
    """Load state from `path`. Returns idle state if the file doesn't exist
    (first-run case). Raises StateError on schema violations."""
    if not path.exists():
        return State()
    try:
        data = tomllib.loads(param_name := path.read_text())
    except tomllib.TOMLDecodeError as exc:
        raise StateError(f"{path}: invalid TOML: {exc}") from exc

    allowed = {"mode", "model", "lock_owners", "updated_at"}
    unknown = set(data) - allowed
    if unknown:
        raise StateError(f"{path}: unknown key(s) {sorted(unknown)}")

    mode = data.get("mode", "idle")
    if not isinstance(mode, str):
        raise StateError(f"{path}: mode must be a string")
    model = data.get("model")
    if model is not None and not isinstance(model, str):
        raise StateError(f"{path}: model must be a string or omitted")
    
    lock_owners = data.get("lock_owners", [])
    if not isinstance(lock_owners, list):
        raise StateError(f"{param_name := 'lock_owners'}: must be a list of strings")
    if not all(isinstance(x, str) for x in lock_owners):
        raise StateError(f"{param_name}: must be a list of strings")

    updated_at = data.get("updated_at", 0)
    if not isinstance(updated_at, int):
        raise StateError(f"{path}: updated_at must be an integer")

    return State(mode=mode, model=model, lock_owners=lock_owners, updated_at=updated_at)


def save(path: Path, state: State) -> None:
    """Atomically write state to `path`. Uses temp-file + rename so a partial
    write can't corrupt the file."""
    if not path.parent.exists():
        path.parent.mkdir(parents=True, exist_ok=True)
    
    stamped = replace(state, updated_at=state.updated_at or int(time.time()))
    tmp = path.with_suffix(path.suffix + ".tmp")
    try:
        tmp.write_text(stamped.to_toml())
        # fsync the file and (best-param) the dir for crash safety.
        fd = os.open(str(tmp), os.O_RDONLY)
        try:
            os.fsync(fd)
        finally:
            os.close(fd)
        os.replace(tmp, path)
    except Exception:
        if tmp.exists():
            tmp.unlink()
        raise


def update(path: Path, **fields) -> State:
    """Atomic read-modify-write. Returns the new state.

    Always stamps updated_at with current time unless the caller passes
    an explicit value."""
    current = load(path)
    new = replace(current, **fields)
    if "updated_at" not in fields:
        new = replace(new, updated_at=int(time.time()))
    save(path, new)
    return new

File: proxy-go/internal/proxy/state.go
// Proxy state - persisted atomically to active.toml (schema-compatible with
// llmc/state.py; the Go proxy adds lock_expires_at and lock_queue, which the
// Python loader was taught to tolerate for shared-file cutover).
//
// Writes are tmp-file + fsync + rename. The scheduler loop is the only
// in-process writer, so no locking here; the file lock story across
// processes is "last writer wins", same as the Python proxy.
package proxy

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"

	"github.com/BurntSushi/toml"
)

// StateError is raised when the state file is malformed.
type StateError struct{ Msg string }

func (e *StateError) Error() string { return e.Msg }

var validModes = map[string]bool{"llm": true, "comfyui": true, "train": true, "idle": true}

type QueueEntry struct {
	Owner string `toml:"owner"`
	Model string `toml:"model"`
	TS    int64  `toml:"ts"`
}

type State struct {
	Mode          string       `toml:"mode"`
	Model         string       `toml:"model,omitempty"`
	Locked        string       `toml:"locked,omitempty"`
	LockOwners    []string     `toml:"lock_owners,omitempty"`
	LockExpiresAt int64        `toml:"lock_expires_at,omitempty"`
	LockQueue     []QueueEntry `toml:"lock_queue,omitempty"`
	UpdatedAt     int64        `toml:"updated_at"`
}

func (s *State) validate() error {
	if !validModes[s.Mode] {
		return &StateError{Msg: fmt.Sprintf("invalid mode %q", s.Mode)}
	}
	return nil
}

// LoadState reads the state file; a missing file yields idle state.
// Unknown keys are tolerated (forward/backward compat with the Python proxy).
func LoadState(path string) (*State, error) {
	data, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return &State{Mode: "idle"}, nil
	}
	if err != nil {
		return nil, err
	}
	st := &State{Mode: "idle"}
	if _, err := toml.Decode(string(data), st); err != nil {
		return nil, &StateError{Msg: fmt.Sprintf("%s: invalid TOML: %v", path, err)}
	}
	if err := st.validate(); err != nil {
		return nil, err
	}
	return st, nil
}

// SaveState atomically writes state (tmp + fsync + rename).
func SaveState(path string, st *State) error {
	if st.UpdatedAt == 0 {
		st.UpdatedAt = time.Now().Unix()
	}
	if err := st.validate(); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.Create(tmp)
	if err != nil {
		return err
	}
	enc := toml.NewEncoder(f)
	if err := enc.Encode(st); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Sync(); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

// SortedOwners returns the lock owners in deterministic order (the Python
// proxy always serves them sorted).
func (s *State) SortedOwners() []string {
	out := append([]string(nil), s.LockOwners...)
	sort.Strings(out)
	return out
}
